### ISSUE TITLE
Add <meta> properties for social media sharing preview images

### ISSUE THREAD

**Issue by aksajhfduwafe** [aksajhfduwafe/bugzilla-to-github-issues-test#5312]
| Field | Value |
|---|---|
| **Reporter** | @axeluhl |
| **Assignee** | @daulenba |
| **Product** | Sailing Race Analytics |
| **Component** | Frontend |
| **Version** | unspecified |
| **Hardware** | All |
| **OS** | All |
| **Severity** | normal |
| **Priority** | P1 |
| **CC (unmapped)** | Lisa Bühn (`lisa.buehn@sap.com`), Marcus Baur (`marcus.baur@sapsailing.com`), Steffen Schäfer (`steffen.schaefer@pgt.de`) |
| **Created** | 2020-06-30T10:43:33Z |
| **Original Bug** | [Bug 5312](https://bugzilla.sapsailing.com/bugzilla/show_bug.cgi?id=5312) |

---

See, e.g., https://css-tricks.com/essential-meta-tags-social-media/ and https://ogp.me/ and https://webapps.stackexchange.com/questions/91864/how-can-i-change-the-preview-image-when-sharing-a-link-on-facebook.

I suggest using the teaser image, if provided, as a preview image, and the event title/description for the title/description properties.

**Comment by aksajhfduwafe** [comment]
**@axeluhl** commented:

(In reply to Dennis Aulenbacher from comment 8 (on this issue))
> In correspondence with Georg we came up with a potential solution: 
> 
> The "share" feature generates specific sharing URLs, that contain the
> parameters needed as path variables. 
> These are resolved by the server and as a response we send a HTML document
> with the meta tags in the header, as well as a meta tag issuing a redirect. 
> The redirect then leads to the original GWT Places page. 
> 
> Example shared URL would look something like this: 
> sapsailing.com/events/dff7d969-9858-4dff-b2f2-a45018a2f755/regatta/
> SAP%20Opti-Cup%202015%20-%20Opti%20A
> 
> @Axel would this be an acceptable URL pattern, or would you prefer a prefix
> like "/shared/"

The pattern is generally okay. However, I don't get how it works. The meta tag causing the redirect would lead to a replacement of the URL in the browser, wouldn't it? So sharing that URL later straight from the browser wouldn't lead to the correct event-specific meta-tags for the media snippets, would it?

> This way we avoid rewriting the URL in the SPA after the User has landed

What does "SPA" stand for?

> there or keeping stale meta tags on the site. 
> Further, we do not need to touch the existing pages / documents at all. 

That would be a benefit for sure, but I still don't get how this would work? Or wouldn't it work for sharing straight from the browser's URL bar?

> As a side note we should consider, whether there are "private"
> events/regattas etc., that should not be viewable via those meta tags. We
> would expose information about title, description and teaser image before
> the redirect, after which an unauthenticated client would get blocked. Maybe
> block any public sharing of events, that are not publicly available?

Good point; when the server patches the meta tags into the page it has to verify that the user has READ permission for any EVENT / REGATTA object affected.

**Comment by aksajhfduwafe** [comment]
**@daulenba** commented:

(In reply to Axel Uhl from comment 9 (on this issue))
> The pattern is generally okay. However, I don't get how it works. The meta
> tag causing the redirect would lead to a replacement of the URL in the
> browser, wouldn't it? So sharing that URL later straight from the browser
> wouldn't lead to the correct event-specific meta-tags for the media
> snippets, would it?

Correct. This would only be supported by the designated sharing buttons in the UI. These can be found on the right side of the EventHeader in Home. I would assume such buttons do also exist in the app. If the user tries to share the URL from the Adress bar, the server would still only deliver the plain Home.html.

> What does "SPA" stand for?

I was referring to the single page application.

> That would be a benefit for sure, but I still don't get how this would work?
> Or wouldn't it work for sharing straight from the browser's URL bar?

Exactly. If this is a hard requirement, we will need to think of another solution. Keeping permanent query parameters updated across all pages maybe. Though that potentially pollutes all URLs leading to an event, series or regatta, since they would carry duplicate information (from the query parameters and the place parameters).

> Good point; when the server patches the meta tags into the page it has to
> verify that the user has READ permission for any EVENT / REGATTA object
> affected.

Problem here is: 
The social media pages use specific agents to fetch the information provided by the tags. These agents will be the actual "users" that need to read the open graph tags. Maybe the chance of randomly guessing an eventID is slim enough to trust into the fact, that an authenticated user generating the URL knows it will be visible to the public. 

Or we disable the sharing feature completely for events flagged as "private". Being aware, that this does not do the complex permission system justice.

**Comment by aksajhfduwafe** [comment]
**@axeluhl** commented:

(In reply to Dennis Aulenbacher from comment 10 (on this issue))
> > That would be a benefit for sure, but I still don't get how this would work?
> > Or wouldn't it work for sharing straight from the browser's URL bar?
> 
> Exactly. If this is a hard requirement, we will need to think of another
> solution. Keeping permanent query parameters updated across all pages maybe.
> Though that potentially pollutes all URLs leading to an event, series or
> regatta, since they would carry duplicate information (from the query
> parameters and the place parameters).

Given the timing I think a first version that works for explicit sharing is better than nothing. Certain nly not a hard requirement.

> Or we disable the sharing feature completely for events flagged as
> "private". Being aware, that this does not do the complex permission system
> justice.

Same thing. If such an agent cannot be authenticated as a specific user, the event needs to be readable by the ANY user, in other words be public. So why not use the existing means of authenticating and authorizing?

**Comment by aksajhfduwafe** [comment]
**@daulenba** commented:

First iteration of tier 1 is building on hudson and will be tested on a dev deployment as soon as it is ready. 

To finish off tier one, a definitive default image has to be added and either be bound to a static URL 
or dynamically loaded from the server. Currently I just used the static URL for the 505 Event 2014 on sapsailing.com.
Dominik said he would be able to provide a suitable image within the next week. 

Another point is the default description and title for the website. 
Is there a marketing-approved text that I could use?

I added the static tags to: 
java/com.sap.sailing.gwt.ui/AdminConsole.html
java/com.sap.sailing.gwt.ui/AutoPlay.html
java/com.sap.sailing.gwt.ui/DataMining.html
java/com.sap.sailing.gwt.ui/Home.html
java/com.sap.sailing.gwt.ui/HomeDesktop.html
java/com.sap.sailing.gwt.ui/HomeMobile.html
java/com.sap.sailing.gwt.ui/Leaderboard.html
java/com.sap.sailing.gwt.ui/LeaderboardEditing.html
java/com.sap.sailing.gwt.ui/PairingList.html
java/com.sap.sailing.gwt.ui/RaceBoard.html
java/com.sap.sailing.gwt.ui/RegattaOverview.html
java/com.sap.sailing.gwt.ui/Simulator.html
java/com.sap.sailing.gwt.ui/Spectator.html

I left out:
/com.sap.sailing.gwt.ui/Calendar.html
/com.sap.sailing.gwt.ui/EmbeddedMapAndWindChart.html
/com.sap.sailing.gwt.ui/VideoPopup.html
/com.sap.sailing.gwt.ui/YoutubePopup.html

Does this cover all relevant pages?
